### PR TITLE
added feed and live command with upload.

### LINKED ISSUE #156: add commands `zbox feed` and `zbox live`
see detail on https://0chain.slack.com/archives/C026YA5EYH3/p1639437096347900?thread_ts=1639368472.338600&cid=C026YA5EYH3

- zbox live
- zbox feed

### PR DESCRIPTION
**Pull request by princeparmar**
added feed and live command
feed = upload --sync (live upload of remote file)
live = upload --live (live upload of local file)
Signed-off-by: Pradip Parmar <pradip@alfa-creator.com>

Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/zboxcli/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR

Associated PRs (Link as appropriate):
...

### CODE REVIEW

File: cmd/upload.go [hunk 1]
```diff
@@ -145,6 +145,160 @@ var uploadCmd = &cobra.Command{
 	},
 }
 
+// feedCmd represents upload command with --sync flag
+var feedCmd = &cobra.Command{
+	Use:   "feed",
+	Short: "live upload from remote device",
```
Comment (cnlangzi) [inline cmd/upload.go]: please check https://github.com/0chain/zboxcli#upload , update description.

**Review by cnlangzi — CHANGES_REQUESTED** [review]
please update README.md too

File: README.md [hunk 2]
```diff
@@ -766,6 +768,94 @@ Use `upload --live` to capture video and audio streaming form local devices, and
...
+<details>
+  <summary>live</summary>
+
+![image](https://github.com/0chain/blobber/wiki/uml/usecase/live_upload_live.png)
+
+</details>
+
+## Feed
+
+Use `feed` command to automatically download segment files from remove live feed with `--downloader-args "-f 22"`
```
Comment (cnlangzi) [inline README.md]:
```suggestion
Use `feed` command to automatically download segment files from remote live feed with `--downloader-args "-q -f best"`
```

File: README.md [hunk 3]
```diff
@@ -766,6 +768,94 @@ Use `upload --live` to capture video and audio streaming form local devices, and
...
+Use `live` to capture video and audio streaming form local devices, and upload
+
+The user must be the owner of the allocation.You can request the file be encrypted before upload, and can send thumbnails with the file. 
+
+| Parameter               | Required | Description                                          | Default | Valid values                            |
+|-------------------------|----------|------------------------------------------------------|---------|-----------------------------------------|
+| allocation              | yes      | allocation id, sender must be allocation owner                     |         | string                 |
+| commit                  | no       | save metadata to blockchain                                        | false   | boolean                |
+| encrypt                 | no       | encrypt file before upload                                         | false   | boolean                |
+| localpath               | yes      | local path of the file to upload                                   |         | file path              |
```
Comment (cnlangzi) [inline README.md]:
```suggestion
| localpath               | yes      | local path of segment files to download, generate and upload                                   |         | file path              |
```

File: README.md [hunk 4]
```diff
@@ -766,6 +768,94 @@ Use `upload --live` to capture video and audio streaming form local devices, and
...
+ - please use `youtube-dl -F https://www.youtube.com/watch?v=pC5mGB5enkw` to list formats of video (see below).
+
+The user must be the owner of the allocation.You can request the file be encrypted before upload, and can send thumbnails with the file. 
+
+| Parameter               | Required | Description                                          | Default | Valid values                            |
+|-------------------------|----------|------------------------------------------------------|---------|-----------------------------------------|
+| allocation              | yes      | allocation id, sender must be allocation owner                     |         | string                 |
+| commit                  | no       | save metadata to blockchain                                        | false   | boolean                |
+| encrypt                 | no       | encrypt file before upload                                         | false   | boolean                |
+| localpath               | yes      | local path of the file to upload                                   |         | file path              |
```
Comment (cnlangzi) [inline README.md]:
```suggestion
| localpath               | yes      | local path of segment files to download, generate and upload                                   |         | file path              |
```

File: cmd/upload.go (hunk 1 above)
Comment (cnlangzi) [inline cmd/upload.go]:
```suggestion
	Short: "download segment files from remote live feed, and upload",
```

File: cmd/upload.go [hunk 5]
```diff
@@ -145,6 +145,160 @@ var uploadCmd = &cobra.Command{
...
+	Long:  `live upload from remote device`,
```
Comment (cnlangzi) [inline cmd/upload.go]:
```suggestion
	Long: "download segment files from remote live feed, and upload",
```

File: cmd/upload.go [hunk 6]
```diff
@@ -145,6 +145,160 @@ var uploadCmd = &cobra.Command{
...
+			commitMetaTxn(remotepath, "Upload", "", "", allocationObj, nil, statusBar)
+			statusBar.wg.Wait()
+		}
+	},
+}
+
+// liveCmd represents upload command with --live flag
+var liveCmd = &cobra.Command{
+	Use:   "live",
+	Short: "live upload from local device",
```
Comment (cnlangzi) [inline cmd/upload.go]:
```suggestion
	Short: "capture video and audio streaming form local devices, and upload",
```

File: cmd/upload.go [hunk 7]
```diff
@@ -145,6 +145,160 @@ var uploadCmd = &cobra.Command{
...
+	Long:  `live upload from local device`,
```
Comment (cnlangzi) [inline cmd/upload.go]:
```suggestion
	Long: "capture video and audio streaming form local devices, and upload",
```

**Review by cnlangzi — CHANGES_REQUESTED** [review]

File: cmd/upload.go [hunk 8]
```diff
@@ -113,19 +113,163 @@ var uploadCmd = &cobra.Command{
...
-			// download video from remote live feed(eg youtube), and sync it to zcn
-			err = startSyncUpload(cmd, allocationObj, localpath, remotepath, encrypt, chunkSize, attrs)
-		} else {
-			err = startChunkedUpload(cmd, allocationObj, localpath, thumbnailpath, remotepath, encrypt, chunkSize, attrs, statusBar, false)
+		err = startChunkedUpload(cmd, allocationObj, localpath, thumbnailpath, remotepath, encrypt, chunkSize, attrs, statusBar, false)
+
+		if err != nil {
+			PrintError("Upload failed.", err)
+			os.Exit(1)
+		}
```
Comment (vincenzopalazzo) [inline cmd/upload.go]:
```suggestion
		if err := startChunkedUpload(cmd, allocationObj, localpath, thumbnailpath, remotepath, encrypt, chunkSize, attrs, statusBar, false); err != nil {
			PrintError("Upload failed.", err)
			os.Exit(1)
		}
```

File: cmd/upload.go [hunk 9]
```diff
@@ -113,19 +113,163 @@ var uploadCmd = &cobra.Command{
...
+			if wps, err = fflags.GetString("attr-who-pays-for-reads"); err != nil {
+				log.Fatalf("getting 'attr-who-pays-for-reads' flag: %v", err)
+			}
+			if err = wp.Parse(wps); err != nil {
+				log.Fatal(err)
+			}
+			attrs.WhoPaysForReads = wp // set given value
+		}
+
+		chunkSize, _ := cmd.Flags().GetInt("chunksize")
```
Comment (vincenzopalazzo) [inline cmd/upload.go:190]:
```suggestion
		chunkSize, err := cmd.Flags().GetInt("chunksize")
		if err != nil {
		  log.Debug(err)
		}
```

More information are always better

File: cmd/upload.go [hunk 10]
```diff
@@ -113,19 +113,163 @@ var uploadCmd = &cobra.Command{
...
+		wg.Add(1)
+		if strings.HasPrefix(remotepath, "/Encrypted") {
+			encrypt = true
+		}
+		var attrs fileref.Attributes
+		if fflags.Changed("attr-who-pays-for-reads") {
+			var (
+				wp  common.WhoPays
+				wps string
+			)
```
Comment (vincenzopalazzo) [inline cmd/upload.go]:
```suggestion
	var wp common.WhoPays
	var wps string
```

IMO this is an antipattern in Go lang

**Review by vincenzopalazzo — APPROVED** [review]
LGTM, just a couple of stylistic comment that you can ignore

**Review by cnlangzi — APPROVED** [review]

**Review by cnlangzi — CHANGES_REQUESTED** [review]
please remove `--live` and `--sync` tags from `upload` command. and update README.md, system_test and 0proxy. thanks.

**Review by cnlangzi — APPROVED** [review]

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feature/156-live-feed-command  |
| gosdk  | staging  |
| zboxcli  | feature/156-live-feed-command  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**